proto/minder/v1/minder.proto

```diff
@@ -4024,6 +4028,188 @@ message EntityInstance {
+    // results is the list of entities
+    repeated EntityInstance results = 1 [
+        (google.api.field_behavior) = REQUIRED
+    ];
+
+    // page is the pagination information
+    CursorPage page = 2;
+}
+
+// GetEntityByIdRequest is the request message for the GetEntityById method
+message GetEntityByIdRequest {
+    // context is the context in which the entity is evaluated
+    ContextV2 context = 1;
+
+    // id is the ID of the entity to get
+    string id = 2 [
+        (buf.validate.field).string = {uuid: true},
+        (google.api.field_behavior) = REQUIRED
+    ];
+}
+
+// GetEntityByIdResponse is the response message for the GetEntityById method
+message GetEntityByIdResponse {
+    // entity is the entity that was retrieved
+    EntityInstance entity = 1 [
+        (google.api.field_behavior) = REQUIRED
+    ];
+}
+
+// GetEntityByNameRequest is the request message for the GetEntityByName method
+message GetEntityByNameRequest {
+    // context is the context in which the entity is evaluated
+    ContextV2 context = 1;
+
+    // name is the name of the entity to get
+    string name = 2 [
+        (buf.validate.field).string = {
+            pattern: "^[A-Za-z][-/[:word:]]*$",
+            max_len: 200,
+        },
+        (google.api.field_behavior) = REQUIRED
+    ];
+
+    // entity_type is the type of entity to get
+    Entity entity_type = 3 [
+        (google.api.field_behavior) = REQUIRED
+    ];
+}
+
+// GetEntityByNameResponse is the response message for the GetEntityByName method
+message GetEntityByNameResponse {
+    // entity is the entity that was retrieved
+    EntityInstance entity = 1 [
+        (google.api.field_behavior) = REQUIRED
+    ];
+}
+
+// DeleteEntityByIdRequest is the request message for the DeleteEntityById method
+message DeleteEntityByIdRequest {
+    // context is the context in which the entity is evaluated
+    ContextV2 context = 1;
+
+    // id is the ID of the entity to delete
+    string id = 2 [
+        (buf.validate.field).string = {uuid: true},
+        (google.api.field_behavior) = REQUIRED
+    ];
+}
+
+// DeleteEntityByIdResponse is the response message for the DeleteEntityById method
+message DeleteEntityByIdResponse {
+    // id is the ID of the entity that was deleted
+    string id = 1 [
+        (google.api.field_behavior) = REQUIRED
+    ];
+}
+
+// RegisterEntityRequest is the request message for the RegisterEntity method
+message RegisterEntityRequest {
+    // context is the context in which the entity is created
+    ContextV2 context = 1;
+
+    // entity_type is the type of entity to create
+    Entity entity_type = 2 [
+        (google.api.field_behavior) = REQUIRED
+    ];
+
+    // identifying_properties is a map of the minimal set of properties
+    // the provider needs in order to uniquely identify an entity.
+    // In most cases, upstream_id will suffice, but this is really
+    // provider-dependent.
+    google.protobuf.Struct identifying_properties = 3;
```

Comment:
This feels "weird" to me.  Specifying a type and a "name in the provider context" feels more natural and less surprising to me.  Consider the following request:

```
context:
  project: "ABC123"
  provider: "gitlab"
entity_type: REPOSITORY
identifying_properties:
  map:
    "github/repo_owner": "JAORMX"  # From github repo
    "tag": "fancybranch"  # From github release
    "gitlab/license": "Apache-2.0"  # From gitlab repo
    "gitlab/project_id": "12460"  # From gitlab MR
    "upstream_id": "12460"  # The thing you mention here
```

Given that we don't expect users to be able to set the other properties, taking a full-blown object (not even a string-string map), and then throwing out all the other keys feels like a clumsy interface on both sides.

Reply:
Yeah, it's not ideal that you could pass anything; but the multi-provider nature of this makes it kinda complicated. Ideally it should look as follows:

```
identifying_properties:
  map:
    "upstream_id": "12460"  # The thing you mention here
```

Since you only need to pass the bare minimum and the rest of the properties will be ignored. This works in _some_ providers, but GitHub... again... is kinda special since it does need the owner and name to properly identify a resource.

Reply:
For GitHub repos, you could use the `owner/repo` slug, which is common enough that it works on the gh CLI and elsewhere.

I feel like there could be a "list unregistered" that gives strings in a provider-specific format, and then a "register" that takes strings in the same format.

Reply:
So, you think we should replace the github upstream_id to leverage that slug?

Reply:
For input from users, the slug feels correct and friendly.  For storage, the numeric upstream id is correct to support repo renames and moves.  For output, we should include both.

I'm not particularly worried about a race between a user supplying a slug in registration and our looking up the id through the provider.  I *do* think we need to hook through the provider during the create, which will give the opportunity to resolve the upstream id and also to register webhooks, etc where needed.